firmware/controllers/actuators/electronic_throttle.cpp

```diff
@@ -264,11 +269,33 @@ static percent_t currentEtbDuty;
 		etbPid.iTermMin = engineConfiguration->etb_iTermMin;
 		etbPid.iTermMax = engineConfiguration->etb_iTermMax;
 
-		currentEtbDuty = engine->engineState.etbFeedForward +
-				etbPid.getOutput(targetPosition, actualThrottlePosition);
+		currentEtbDuty = engine->engineState.etbFeedForward + etbPid.getOutput(targetPosition, actualThrottlePosition);
 
 		etb1.dcMotor.Set(ETB_PERCENT_TO_DUTY(currentEtbDuty));
 
+	   	float etbError = absF(targetPosition - actualThrottlePosition); // Used for I-Term limiter and Error-detection
+
+
+        // ETB Deviation monitoring
+        if (etbError < 3 ) {
+          start_time = currentTimeMillis(); //Updates timestamp as long as ETB-Error stays within 3% deviation
+        }
+      else {
+        }        // ETB Deviation outside limit detected, set timestamp and monitor it for two seconds
+        if (chVTIsSystemTimeWithinX(start_time, start_time + TIME_MS2I(2000))) {
+        	  //Less than two seconds, system OK: do nothing
+            }
+       else {
+    	   if (rpm > engineConfiguration->cranking.rpm) {  //We dont want or need any of this to execute if engine are not running
+                         // It would make a huge issue while tuning ETB
+        	   etbDeviationError = true;
+        	   warning(OBD_Throttle_Actuator_Control_Range_Performance_Bank_1, "Electronic throttle deviation limits exceeded"); //Set Error-code
+        	  // engineConfiguration->rpmHardLimit = 2000;  //Set RPM Limit to 2000 to not kill engine if ETB opening are too big
+        	   etb1.dcMotor.Set(0); // Disable ETB Duty
```

Comment:
this won't actually disable the throttle - it'll continue to operate, since just after this call we'll just reset it with the PID'd duty

Reply:
Perfect catch.

We need automated test coverage. I am sorry but I cannot rely on humans promising me that their code works :(

Reply:
let's just fire the warning for now - like I mentioned on the other comment, we need somewhat more comprehensive limp home, which needs to be designed properly instead of tacking in things like "oh and here we disable the ETB motor"

Reply:
How long does that take?

I am having it shut down in error now, but nothing happens..

Reply:
It sort of doesn't matter how long it takes.  The ETB is safety-critical hardware, so we really shouldn't do anything half-baked or "hackey" when dealing with it.

Reply:
That's the whole point of this function.

This is the proper way to monitor and catch all etb-issues.

I am not a professional programmer, that's why I have asked if you can help refining it

Reply:
And the right reaction for this error are to disable etb, to not fry anything + set the rpm limit to not blow your engine.

Reply:
A "Limp mode" would be a additional thing, but are never used in a aftermarket

Reply:
There are 2.5 sides to the desired behavior here:
1) detect a problem - this code does a good job of that
2)
    a) report the problem to the user
    b) do something safe with the engine/hardware/car/whatever

This PR does a fine job of 1 and 2a.  2b is a larger problem that requires involvement of many systems other than just the electronic throttle controller.  Let's deal with 1 and 2a in this PR, and then figure out the right way to tackle 2b.

Reply:
By the way, disabling ETB, setting a rev limit, etc are absolutely a limp mode.  Call it what you want, but everything in the realm of "impose lower limits to prevent engine damage" is limp mode.

Reply:
Logically separating those concerns helps improve code quality and reliability.  Which for this particular case means:
- Detecting an error when there is one (prevent false negatives)
- NOT detecting an error when there isn't one (prevent false positives)

Both may be equally dangerous if they can alter engine behavior.

Reply:
In 99% of the cases, people go aftermarket to avoid the limp-mode who are a issue when tuning a OEM.
For this specific one, I would not call it limp-mode.
If you have deviation over time at the etb, something are seriously wrong, and it would probably shut down soon because of overtemp.

But no reason to discuss this.
At least the function works as it should, and does a good job. I am sure we find out the last step.

Reply:
Just tested with directPwmValue = 0; instead now, and it's stone dead even after error are corrected. 
Only way are reset, just like it should.

I used that for the Tps Calibration, and could not get it back running after calibration.
Tried everything, and the only thing that could get it moving again was directPwmValue = NAN; or reset